index.html

```diff
@@ -3930,6 +3938,64 @@ <h2>Identification</h2>
   CoAP Content-Format ID <code>432</code> (see <a href="#iana-section"></a>).
   </p>
 
+</section>
+
+<section id="canonicalization-serialization-json">
+<h2>Canonicalization</h2>
+
+  <p>
+  Canonicalization is a process applied to the JSON serialization of a TD that
+  removes variations between different expressions of the same information model,
+  so that the same information always results in the same JSON serialization.
+  This unique representation is called a <a>Canonical TD</a>.
+  Canonical TDs are important for applications that need to compare TDs as strings,
+  for example, to compute signatures on TDs or to compute differences between TDs.
+  Whenever there is a choice of multiple ways to present information, canonicalization
+  selects only one. There are many places in which optional features are included in JSON,
+  such as white space, or where different representation formats are possible, such as
+  the number of digits in a number representation or the order of elements in an object.
+  </p><p>
+  Some of the choices leading to the efficient computation of a canonical form, 
+  such as the suppression of white space and sorting of elements in objects 
+  do make the representation less human readable.
+  In general, the canonical form presented here is optimized for machine-to-machine
+  use cases such as signing and to accommodate devices with limited network and
+  compute capabilities.  Compactness and ease of implementation
+  has been emphasized over human readability.
+  </p>
+  <p>Computing a canonical form directly from an RDF triple representation is
+  difficult so the starting point is the JSON serialization of a TD.  
+  The following sequence of transformations are then applied to this serialization: 
+  <ol>
+  <li>Remove any elements that are given their default value. 
+  For example, the default value of <code>observable</code>
+  is <code>false</code>.  
+  If an input TD explictly assigns the value <code>false</code> to <code>observable</code>,
+  this element and the redundant value will be removed.</li>
+  <li>Rewrite values that can be expressed as either an array or as a single value 
+  as a single value if there is only one element.  
+  In other words, remove square brackets around
+  arrays of single elements where this is permitted.
+  </li>
```

Comment:
It will be a programming burden to process a defaults list and remove attributes from a complex object. What about extensions?
I would have suggested only removing `null` assignments if I were sure there are no domain-specific semantics behind them.

Reply:
The problem is that when you pull things into a database, you will fill in all the default values.  Later you don't know whether a value was assigned during import or by the originator.  Would only apply to defaults defined in the TD spec, not in extensions.

Reply:
Why doesn't it apply to extensions? Extensions can also default values and the same issue arises when pulling that data out of such default-aware databases. The canonicalization algorithm should work on extended TDs too. 

I think this will be easier if TD spec provides the defaults in a simple machine-readable and extensible way. Maybe a TD with only default values to compare against the input TD during this stage of the algorithm?

Reply:
Defaults are handled in a special way by a TD processor, basically filling in missing items during import.  Extensions would not have defaults; we might even want to require this.

Reply:
I think they would have (unless we are talking about different things). Some examples:
https://www.w3.org/TR/wot-thing-description11/#http-binding-assertions
https://www.w3.org/TR/wot-binding-templates/#mqtt-default-vocabulary-terms